Mathlib/NumberTheory/ModularForms/JacobiTheta/TwoVariable.lean

```diff
@@ -14,119 +15,501 @@ This file defines the two-variable Jacobi theta function
+  rw [jacobiTheta₂', jacobiTheta₂', ← tsum_neg, ← (Equiv.neg ℤ).tsum_eq]
+  congr 1 with n
+  simp only [jacobiTheta₂'_term, jacobiTheta₂_term]
+  rw [Equiv.neg_apply, ← neg_mul]
+  push_cast
+  ring_nf
+
+lemma jacobiTheta₂'_conj (z τ : ℂ) :
+    conj (jacobiTheta₂' z τ) = jacobiTheta₂' (conj z) (-conj τ) := by
+  rw [← neg_inj, ← jacobiTheta₂'_neg_left, jacobiTheta₂', jacobiTheta₂', conj_tsum, ← tsum_neg]
+  congr 1 with n
+  simp_rw [jacobiTheta₂'_term, jacobiTheta₂_term, map_mul, ← Complex.exp_conj, map_add, map_mul,
+    ← ofReal_int_cast,← ofReal_ofNat, map_pow, conj_ofReal, conj_I]
+  ring_nf
+
+/-!
+## Functional equations
+-/
+
 /-- The functional equation for the Jacobi theta function: `jacobiTheta₂ x τ` is an explict factor
 times `jacobiTheta₂ (x / τ) (-1 / τ)`. This is the key lemma behind the proof of the functional
-equation for Dirichlet L-series. -/
-theorem jacobiTheta₂_functional_equation (z : ℂ) {τ : ℂ} (hτ : 0 < im τ) :
-    jacobiTheta₂ z τ =
-      1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (-π * I * z ^ 2 / τ) * jacobiTheta₂ (z / τ) (-1 / τ) := by
+equation for L-series of even Dirichlet characters. -/
+theorem jacobiTheta₂_functional_equation (z τ : ℂ) : jacobiTheta₂ z τ =
+    1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (-π * I * z ^ 2 / τ) * jacobiTheta₂ (z / τ) (-1 / τ) := by
+  rcases le_or_lt (im τ) 0 with hτ | hτ
+  · rw [jacobiTheta₂_undef z hτ, jacobiTheta₂_undef, mul_zero]
+    rw [neg_div, neg_im, one_div, inv_im, neg_nonpos]
+    exact div_nonneg (neg_nonneg.mpr hτ) (normSq_nonneg τ)
+  unfold jacobiTheta₂ jacobiTheta₂_term
   have h0 : τ ≠ 0 := by contrapose! hτ; rw [hτ, zero_im]
   have h2 : 0 < (-I * τ).re := by
     simpa only [neg_mul, neg_re, mul_re, I_re, zero_mul, I_im, one_mul, zero_sub, neg_neg] using hτ
   calc
-    _ = ∑' n : ℤ, cexp (-π * (-I * τ) * ↑n ^ 2 + 2 * π * (I * z) * ↑n) :=
-      tsum_congr (fun n ↦ by ring_nf)
-    _ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * ∑' (n : ℤ), cexp (-π / (-I * τ) * (n + I * (I * z)) ^ 2) := by
-      rw [tsum_exp_neg_quadratic h2]
-    _ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (π * I * (-1 / τ) * z ^ 2) *
-        ∑' (n : ℤ), cexp (2 * π * I * n * (z / τ) + π * I * n ^ 2 * (-1 / τ)) := by
-      simp_rw [mul_assoc _ (cexp _), ← tsum_mul_left (a := cexp _), ← Complex.exp_add]
-      congr 2 with n : 1; congr 1
-      field_simp [I_ne_zero]
-      ring_nf
-      simp_rw [I_sq, I_pow_four]
-      ring_nf
-    _ = _ := by
-      congr 3
-      ring
+  _ = ∑' n : ℤ, cexp (-π * (-I * τ) * ↑n ^ 2 + 2 * π * (I * z) * ↑n) :=
+    tsum_congr (fun n ↦ by ring_nf)
+  _ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * ∑' (n : ℤ), cexp (-π / (-I * τ) * (n + I * (I * z)) ^ 2) := by
+    rw [tsum_exp_neg_quadratic h2]
+  _ = 1 / (-I * τ) ^ (1 / 2 : ℂ) * cexp (π * I * (-1 / τ) * z ^ 2) *
+      ∑' (n : ℤ), cexp (2 * π * I * n * (z / τ) + π * I * n ^ 2 * (-1 / τ)) := by
+    simp_rw [mul_assoc _ (cexp _), ← tsum_mul_left (a := cexp _), ← Complex.exp_add]
+    congr 2 with n : 1; congr 1
+    field_simp [I_ne_zero]
```

Comment:
`field_simp` is slow, but it's a fairly complicated equation, so it's probably OK.